### PR TITLE
Add friendly_name to device

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This adds a new `friendly_name` configuraton option to the core config.

This friendly name will be used as the config entry and device names in Home Assistant and will be automatically prefixed to all entity names in Home Assistant, removing the need to prefix them all with a substitution in the YAML files.

### Notes
No work has been done on the mqtt side to use this yet.

### Related PRs:
- https://github.com/esphome/dashboard/pull/416
- https://github.com/esphome/aioesphomeapi/pull/362
- https://github.com/home-assistant/core/pull/85976

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: my-device
  friendly_name: My Device

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nielsnl68** [comment]
Hi Jesse,
is `device_name` not a beter term than `friendly_name`?

File: esphome/dashboard/util.py [hunk 1]
```diff
@@ -7,3 +10,16 @@ def password_hash(password: str) -> bytes:
...
+    return "".join(
+        c
+        for c in unicodedata.normalize("NFD", str(value))
+        if unicodedata.category(c) != "Mn"
+    )
+
+
+def friendly_name_slugify(value):
+    value = strip_accents(value).lower().replace(" ", "-").replace("_", "-")
+    return "".join(c for c in value if c in ALLOWED_NAME_CHARS)
```
Comment (balloob) [inline esphome/dashboard/util.py:32]: let’s make sure it’s never an empty string

Comment (jesserockz) [reply esphome/dashboard/util.py:32]: This is what is sent from the frontend as `name` to the API, unless the frontend sends empty, it wont be?

Comment (balloob) [reply esphome/dashboard/util.py:32]: What if the frontend sends a name with only characters not in `ALLOWED_NAME_CHARS` ?

**Pull request merged**